web/tests/lib/make_message.js

```diff
@@ -0,0 +1,142 @@
+"use strict";
+
+function gen_random_int(seed) {
+    return Math.floor(Math.random() * seed + 1);
+}
+
+const DEFAULT_REALM_STR = "example";
+
+const stream_id = gen_random_int(1000);
+
+const recipient_id = gen_random_int(1000) + stream_id;
+
+let prev_sender_id = 100;
+let prev_msg_id = 10000;
+
+const DEFAULT_MSG_TYPE = gen_random_int(1000) % 2 === 0 ? "stream" : "private";
+
+function create_sender_id() {
+    prev_sender_id += 1;
+    return prev_sender_id;
+}
+
+function create_msg_id() {
+    prev_msg_id += 1;
+    return prev_msg_id;
+}
+
+function create_display_recip(msg_type = DEFAULT_MSG_TYPE, sender_id) {
+    if (msg_type === "stream") {
+        return `stream_${stream_id}`;
+    }
+
+    const pm_recipient = [];
+    const self_recipient = {
+        id: sender_id,
+        email: `user_${sender_id}@example.com`,
+        full_name: `user_${sender_id}`,
+    };
+    pm_recipient.push(self_recipient);
+
+    return pm_recipient;
+}
+
+exports.create_display_recip = create_display_recip;
+
+function create_stream_id(msg_type) {
+    if (msg_type === "private") {
+        return undefined;
+    }
+
+    return stream_id;
+}
+
+function get_subject(msg_type) {
+    if (msg_type === "private") {
+        // 'subject' is an empty string for private messages.
+        return "";
+    }
+
+    return `example_${stream_id}-${recipient_id}`;
+}
+
+function create_reactions(sender_id) {
+    const reactions = [];
+
+    // for simplicity the reaction object contains fields of sender ie. we are assuming
+    // a self-reaction to the message.
+    const self_react = {
+        emoji_name: "+1",
+        emoji_code: "1f44d",
+        reaction_type: "unicode_emoji",
+        user: {
+            email: `user_${sender_id}@example.com`,
+            id: sender_id,
+            full_name: `user_${sender_id}`,
+        },
+        user_id: sender_id,
+    };
+
+    reactions.push(self_react);
+
+    return reactions;
+}
+
+function create_recipient_id() {
+    return recipient_id;
+}
+
+function create_submessage(msg_id) {
+    const submessages = [];
+
+    submessages.push({
+        msg_type: DEFAULT_MSG_TYPE,
+        content: "<p>(submessage content)</p>",
+        message_id: msg_id,
+        sender_id: gen_random_int(99),
+        id: gen_random_int(100) + msg_id,
+    });
+
+    return submessages;
+}
+
+function create_topic_links() {
+    const topic_links = [];
+
+    topic_links.push({
+        url: `https://example_${stream_id}-${recipient_id}.com`,
+        text: `link_${stream_id}-${recipient_id}`,
+    });
+}
+
+exports.make_server_message = (opts = {}) => {
+    const sender_id = opts.sender_id ?? create_sender_id();
+    const msg_id = opts.id ?? create_msg_id();
+    const msg_type = opts.type ?? DEFAULT_MSG_TYPE;
```

Comment:
Rather than one function that defaults to either stream/channel messages or DMs (or randomly chooses one or the other as default), I recommend having two functions like `make_channel_message` and `make_direct_message`.

A large fraction of tests that need a message are going to care whether it's a DM or a channel message, so they'll need to specify anyway.

And because the types are effectively different for the two kinds of messages, separating the functions will help a lot in being able to type-check the functions and their callers later, when we eventually make TypeScript possible in our tests. Even before we have a type-checker going, writing the code in a way that a type-checker *could* understand is usually a helpful guideline for writing it in a way that human readers can easily understand.

For an example of how to let those two functions share the parts of their code that make sense to share, see zulip-mobile:
https://github.com/zulip/zulip-mobile/blob/2ae8d8e54dcf25ce760206b0f6fc16923b4375b5/src/__tests__/lib/exampleData.js#L402-L528
and in particular here:
```js
  const baseMessage: StreamMessage = {
    ...messagePropertiesBase,
    ...messagePropertiesFromSender(sender),
    ...messagePropertiesFromStream(streamInner),

    content: '<p>This is an example stream message.</p>',
    content_type: 'text/html',
```

Reply:
Created two separate function to handle both (`"stream"` and `"private"`) kind of messages. 

>>> For an example of how to let those two functions share the parts of their code that make sense to share, see zulip-mobile:

Thanks for the linking that resource. The newer version tries to address the feedback.